samples/button_press_cb/src/button.c

```diff
@@ -0,0 +1,191 @@
+/*
+ * Copyright (c) 2023 T-Mobile USA, Inc.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/device.h>
+#include <zephyr/drivers/gpio.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/printk.h>
+
+#include "button.h"
+
+#define SW0_NODE DT_ALIAS(sw0)
+
+static const struct gpio_dt_spec button = GPIO_DT_SPEC_GET_OR(SW0_NODE, gpios, {0});
+static struct gpio_callback button_cb_data;
+static button_press_callback_t registered_callback;
+static struct gpio_callback button_cb_data;
+button_press_state button_state = BUTTON_STATE_DEINIT;
+struct k_timer debounce_timer;
+struct k_timer timeout_timer;
+static int64_t last_button_press_time;
+static int64_t button_press_duration;
+static int64_t max_button_press_timeout;
+
+/* button GPIO interrupt handler */
+static void button_press_handler(const struct device *port, struct gpio_callback *cb,
+				 gpio_port_pins_t pins)
+{
+	k_timer_stop(&debounce_timer);
+	k_timer_start(&debounce_timer, K_MSEC(50), K_NO_WAIT);
+	button_state = BUTTON_STATE_DEBOUNCING;
+}
+
+static void button_timeout_handler(struct k_timer *p_timeout_timer)
+{
+	button_state = BUTTON_STATE_TIMEOUT;
+	if (NULL != registered_callback) {
+		/* saving the last button press time, since k_uptime_delta updates it with current
+		 * system uptime, which is undesired behavior in the context of measuring the real
+		 * button press duration even if timeout has occurred */
+		int64_t save_last_button_press = last_button_press_time;
+
+		button_press_duration = k_uptime_delta(&last_button_press_time);
+		last_button_press_time = save_last_button_press;
+		/* Invoke the registered user callback handler */
+		registered_callback(last_button_press_time, button_press_duration,
+				    BUTTON_PRESS_TIMEOUT);
+	}
+	k_timer_stop(p_timeout_timer);
+}
+
+/* debounce timer handler */
+static void button_debounce_timer_handler(struct k_timer *p_debounce_timer)
```

Comment:
Why are we implementing our own debounce? We should be using the built in debounce mechanism (gpio keys)

Reply:
I did not know we have that one.
As for conforming to the acceptance criteria in the ticket, I believe having 'some; kind of debounce is good for now.
I would love to explore the already existing debounce mechanism and integrate it to this sample, but it's too late now for this to be done today, we need to release this for tomorrow's 1.20